### DISCUSSION TITLE
[Feature]: Loading custom scripts

### DISCUSSION THREAD

**Discussion by phuoccss** [remix-run/remix#2895] [Proposals]
### What is the new or updated feature that you are suggesting?

Allowing adding of external scripts to routes. Can be either by inlining script tag or via a separate file. The API would look similar to `meta`/`links` functions.

### Why should this feature be included?

I'm mainly thinking about JS-light web pages with hints of interactivity e.g. dark mode switch, canvas app, or usage of any imperative JS library where React is not needed.

Currently, there is `<script dangerouslySetInnerHTML>`. However, this is not the most ergonomic due to lack of syntax highlighting, linting & minification.

Alternatively, a separate process can be set up to build external script files, which can be added via `<script src>` . This allows syntax highlighting and minification. However, to further optimize loading speed, we also need to set up filename hashing and add caching header manually.

By making it a first class feature, we can overcome above limitations. This also fits in with Remix's vision of a progressive enhancement framework.

**Comment by laurisbernhart** [comment]
Would love to see better support too! 

I have disabled `<Scripts/>` and I am using just a bit of Vanilla Js.. 
Atm I am running webpack and bundling those scripts into the public folder.

**Comment by mcintyre94** [comment]
Another use case for this is eg. pyodide: https://pyodide.org/en/stable/usage/quickstart.html#complete-example

There you need to load their JS in a `<script>` tag and can then (in the browser) use its functionality. I don't think there's any way to do this in remix currently other than add it to the `<head>` tag in root.tsx and have it load on every page.

**Comment by sergiodxa** [comment]
Right now, instead of using `<script dangerouslySetInnerHTML>` to inline a script, you could use a Resource Route to generate the script content, then render a script tag with the src pointing to that Resource Route, you could optionally add a cache layer to the response so you don't generate the script every time, and inside the resource route you could minify the script if you want.

**Comment by mcintyre94** [comment]
@sergiodxa Sorry if I'm missing something really obvious, but if you have a resource route that is caching + generating a script, how would you include that script in a single page? [The docs](https://remix.run/docs/en/v1/guides/resource-routes) only talk about linking to them. It's obvious how other use cases like a CSS with `links` or an `<img>` tag pointing at a resource route would work, but I'm not clear how you'd go about adding a script to the `<head>` of a single page.

**Comment by sergiodxa** [comment]
> @sergiodxa Sorry if I'm missing something really obvious, but if you have a resource route that is caching + generating a script, how would you include that script in a single page? [The docs](https://remix.run/docs/en/v1/guides/resource-routes) only talk about linking to them. It's obvious how other use cases like a CSS with `links` or an `<img>` tag pointing at a resource route would work, but I'm not clear how you'd go about adding a script to the `<head>` of a single page.

You can add a script tag with the src pointing to the resource route and render that script in the head.

```html
<script src="/path/to/resource/route" />
```

**Comment by mcintyre94** [comment]
> You can add a script tag with the src pointing to the resource route and render that script in the head.

Thanks, that's definitely what I'd like to do! Is there any documentation of how to add to the head of a single page in remix? In the docs I've only seen `<head>` in `root.tsx`.

Reading the docs a bit more I think I might be able to do it in `root.tsx` using a similar `handle` function to that used in https://remix.run/docs/en/v1/guides/disabling-javascript - except I'd load a particular script in `<head>` if any child pages matched the handle. I think that'd be workable but I'm not sure if it's the best way to do it?

**Comment by sergiodxa** [comment]
@mcintyre94 in Remix Utils I build a component and type to do it https://github.com/sergiodxa/remix-utils#externalscripts

You can also check the code in the repo if you don’t want to add a package, it uses useMatches and the handle export to work.

**Comment by mcintyre94** [comment]
Thanks, that's really helpful! Remix-utils looks cool too, will have to check it out :)

**Comment by rakeshpai** [comment]
It would be great to have the first-class support for this that `links` enjoys. The benefits are:
* As stated above - avoid shipping react to the client and only ship very lightweight JS
* Co-locating scripts with components
* First class support for minification
* Adding / removing script tags like we have for `links` + caching.
* (Possibly) defer scripts
* Typescript support
* Modules support

While it's maybe possible to achieve all the above with ResourceRoutes, it would be clunky and require a lot of boilerplate.
